### PR TITLE
Refactor interpreter `convert()` function

### PR DESCRIPTION
**Pull request by SupunS**
## Description

Move the common `if valueType.Equal(unwrappedTargetType)` check to the top.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by turbolent — COMMENTED** [review]
This is definitely cleaner, but the reasoning for performing the equality tests in the branches was that this avoids unnecessary equality tests for types not handled by the switch.

Benchmarks do not report any difference, but that is probably because we do not have good benchmarks for this code, not tests it seems.

Could you please add some benchmarks to ensure this does not regress performance?

**Comment by SupunS** [comment]
Yeah, that's a fair point. In that case, we can selectively do the type-equality check: d6b546cf2ff068d0eba54081dc4f2ca2e13474c2

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -1620,150 +1620,104 @@ func (interpreter *Interpreter) convert(value Value, valueType, targetType sema.
...
+				unwrappedTargetType.Type,
+			)
 
-			default:
-				panic(errors.NewUnexpectedError("unsupported reference value: %T", ref))
-			}
+		default:
+			panic(errors.NewUnexpectedError("unsupported reference value: %T", ref))
 		}
 	}
```
Comment (turbolent) [inline runtime/interpreter/interpreter.go:1722]: Maybe panic with an unexpected error here: The new check at the top filters which types need conversion – if it allowed a type, and we reached this place, the type's conversion isn't implemented properly (feel free to even add this as a comment)

**Review by turbolent — APPROVED** [review]
Looks good! Just one minor suggestion

File: runtime/interpreter/interpreter.go (hunk 1 above)
Comment (SupunS) [reply runtime/interpreter/interpreter.go:1722]: While adding this, it made me realize that certain values that were allowed from the above check don't need conversion. e.g: Numeric super types.
Now I'm having second thoughts about this check. Having the checks in two places can easily lead to inconsistencies. 
I now think it maybe good to keep it as is.

**Pull request closed without merging**